lavalink/WebSocket.py

```diff
@@ -56,42 +58,71 @@ def __init__(self, lavalink, host, password, ws_port, ws_retry, shard_count):
                 for task in self._queue:
                     await self.send(**task)
 
+    async def _attempt_reconnect(self) -> bool:
+        """
+        Attempts to reconnect to the lavalink server.
+
+        Returns
+        -------
+        bool
+            ``True`` if the reconnection attempt was successful.
+        """
+        await self._ws.close()
+        log.info('Connection closed; attempting to reconnect in 30 seconds')
+        for a in range(0, self._ws_retry):
+            await asyncio.sleep(30)
+            log.info('Reconnecting... (Attempt {})'.format(a + 1))
+            await self.connect()
+
+            if self._ws.open:
+                return True
+        return False
+
     async def listen(self):
-        try:
-            while self._ws.open:
+        while not self._shutdown.is_set():
+            try:
                 data = json.loads(await self._ws.recv())
-                op = data.get('op', None)
-                log.debug('Received websocket data\n' + str(data))
-
-                if not op:
-                    return log.debug('Received websocket message without op\n' + str(data))
-
-                if op == 'event':
-                    await self._lavalink.dispatch_event(
-                        data['type'], data['guildId'], data.get('reason', data['type'])
-                    )
-                elif op == 'playerUpdate':
-                    await self._lavalink.update_state(data)
-        except websockets.ConnectionClosed:
-            self._lavalink.players.clear()
-
-            log.info('Connection closed; attempting to reconnect in 30 seconds')
-            self._ws.close()
-            for a in range(0, self._ws_retry):
-                await asyncio.sleep(30)
-                log.info('Reconnecting... (Attempt {})'.format(a + 1))
-                await self.connect()
+            except websockets.ConnectionClosed:
+                if self._shutdown.is_set():
+                    # Exit gracefully
+                    break
 
-                if self._ws.open:
+                self._lavalink.players.clear()

```

Comment:
This should be moved directly below the `except`, players are cleared by Lavalink anyway when the websocket is disconnected